src/main/java/de/bonndan/nivio/output/map/PNGRenderCache.java

```diff
@@ -0,0 +1,110 @@
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A service that caches expensive batik svg to png rendering.
+ *
+ */
+@Service
+public class PNGRenderCache implements ApplicationListener<ProcessingFinishedEvent> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(PNGRenderCache.class);
+
+    private final Map<String, Pair<LandscapeImpl, byte[]>> renderings = new HashMap<>();
+
+    private final MapFactory<mxGraph, mxCell> mapFactory;
+    private final MapStyleSheetFactory mapStyleSheetFactory;
+
+    public PNGRenderCache(MapFactory<mxGraph, mxCell> mapFactory, MapStyleSheetFactory mapStyleSheetFactory) {
+        this.mapFactory = mapFactory;
+        this.mapStyleSheetFactory = mapStyleSheetFactory;
+    }
+
+    /**
+     * Returns a png.
+     *
+     * @param landscape the landscape to render
+     * @return the png as byte array, cached if available
+     */
+    public byte[] getPNG(LandscapeImpl landscape) {
+
+        Pair<LandscapeImpl, byte[]> landscapePair = renderings.get(landscape.getIdentifier());
+        if (landscapePair == null || landscapePair.getLeft().getLog().getLastUpdate() != landscape.getLog().getLastUpdate()) {
+            byte[] rendered = asByteArray(landscape);
+            renderings.put(landscape.getIdentifier(), new ImmutablePair<>(landscape, rendered));
+            return rendered;
+        }
+
+        return landscapePair.getRight();
+    }
+
+    /**
+     * Returns an svg.
+     *
+     * @param landscape the landscape to render
+     * @return the svg as string, uncached
+     */
+    public String getSVG(LandscapeImpl landscape) {
+        JGraphXRenderer jGraphXRenderer = new JGraphXRenderer();
+        RenderedArtifact<mxGraph, mxCell> render = jGraphXRenderer.render(landscape);
+        mapFactory.applyArtifactValues(landscape, render);
+
+        SvgFactory svgFactory = new SvgFactory(landscape, mapStyleSheetFactory);
+        return svgFactory.getXML();
+    }
+
+    @Override
+    public void onApplicationEvent(ProcessingFinishedEvent processingFinishedEvent) {
+        Landscape landscape = processingFinishedEvent.getLandscape();
+        if (landscape instanceof LandscapeImpl) {
+            LOGGER.info("Generating PNG rendering of landscape {}", landscape.getIdentifier());
+            getPNG((LandscapeImpl) landscape);
+        }
+    }
+
+    private byte[] asByteArray(LandscapeImpl landscape) {
+        TranscoderInput input_svg_image = new TranscoderInput(new ByteArrayInputStream(getSVG(landscape).getBytes()));
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        TranscoderOutput transcoderOutput = new TranscoderOutput(outputStream);
+        Transcoder transcoder = new PNGTranscoder();
+        try {
+            transcoder.transcode(input_svg_image, transcoderOutput);
+        } catch (TranscoderException e) {
+            throw new RuntimeException("Failed to create PNG", e);
+        }
+        try {
```

Comment:
Do you close the outputStream if an Exception occurrs during the flush? Better use try-with-resources for the AutoCloseable OutputStream :) (https://stackoverflow.com/questions/56112598/proper-way-to-close-an-autocloseable)

Reply:
But, for ByteArrayOutputStream, close() and flush() do nothing (check out the implementation). This try-catch-block is not necessary. I wonder now, what the exception message was intended for?